server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+        boolean success = false;
+        try {
+            final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
+            if (snapshotShard != null) {
+                threadPool.executor(executorName()).execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                            logger.debug("skipping snapshot shard size retrieval for {} as node is no longer master", snapshotShard);
+                            return;
+                        }
+
+                        final RepositoriesService repositories = repositoriesService.get();
+                        assert repositories != null;
+                        final Repository repository = repositories.repository(snapshotShard.snapshot.getRepository());
+
+                        logger.debug("fetching snapshot shard size for {}", snapshotShard);
+                        final IndexShardSnapshotStatus status = repository.getShardSnapshotStatus(
+                            snapshotShard.snapshot().getSnapshotId(),
+                            snapshotShard.index(),
+                            snapshotShard.shardId()
+                        );
+
+                        final long snapshotShardSize = status.asCopy().getTotalSize();
+                        logger.debug("snapshot shard size for {}: {} bytes", snapshotShard, snapshotShardSize);
+
+                        boolean updated = false;
+                        synchronized (InternalSnapshotsInfoService.this) {
+                            if (isMaster) {
+                                final ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes =
+                                    ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                                updated = newSnapshotShardSizes.put(snapshotShard, snapshotShardSize) == null;
+                                knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                            }
+                        }
+                        if (updated) {
+                            rerouteService.get().reroute("snapshot shard size updated", Priority.HIGH,
+                                ActionListener.wrap(
+                                    r -> logger.trace("reroute after snapshot shard size update completed"),
+                                    e -> logger.debug("reroute after snapshot shard size update failed", e)));
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn(() -> new ParameterizedMessage("failed to retrieve shard size for {}", snapshotShard), e);
+                    }
+
+                    @Override
+                    public void onAfter() {
+                        unknownSnapshotShards.remove(snapshotShard);
```

Comment:
Should we assert that this returned `true`?

Reply:
I think we can -> a3dcb20796c2dec4189289bd6a7e2fcc554bbf93